components/NewRelicAgent/NRAgent.brs

```diff
@@ -732,6 +732,27 @@ function nrAddAttributes(ev as Object) as Object
     ev.AddReplace("videoMode", dev.GetVideoMode())
     ev.AddReplace("graphicsPlatform", dev.GetGraphicsPlatform())
     ev.AddReplace("timeSinceLastKeypress", dev.TimeSinceLastKeypress() * 1000)
+
+    if isMemoryMonitorAvailable(dev.GetModel())
+        memMonitor = CreateObject("roAppMemoryMonitor")
+
+        'UNDOCUMENTED TRICK:
+        '   If EnableMemoryWarningEvent(true) returns false, a call to GetChannelAvailableMemory will freeze the device on v12.5
+        enableMemWarningRet = memMonitor.EnableMemoryWarningEvent(true)
+        memMonitor.EnableMemoryWarningEvent(false)
+
+        'Available since v10.5
+        if memMonitor <> invalid
+            ev.AddReplace("memLimitPercent", memMonitor.GetMemoryLimitPercent())
+            'Available for RokuOS v12.5+
+            if FindMemberFunction(memMonitor, "GetChannelAvailableMemory") <> Invalid
+                if enableMemWarningRet = true
+                    ev.AddReplace("channelAvailMem", memMonitor.GetChannelAvailableMemory())
+                end if
+            end if
+        end if
+    end if
```

Comment:
Having these checks and allocations on every event will become a bit expensive for the client apps. Would you consider moving most of these checks to initialisation instead?

On init, it would look something like:

```brightscript
if isMemoryMonitorAvailable(dev.GetModel())
    m.enableChannelAvailableMem = false

    'Available since v10.5
    m.memMonitor = CreateObject("roAppMemoryMonitor")
    if m.memMonitor <> invalid
        'UNDOCUMENTED TRICK:
        '   If EnableMemoryWarningEvent(true) returns false, a call to GetChannelAvailableMemory will freeze the device on v12.5
        enableMemWarningRet = m.memMonitor.EnableMemoryWarningEvent(true)
        m.memMonitor.EnableMemoryWarningEvent(false)

        if enableMemWarningRet
            'Available for RokuOS v12.5+
            m.enableChannelAvailableMem = FindMemberFunction(m.memMonitor, "GetChannelAvailableMemory") <> Invalid
        end if
    end if
end if
```

And here, we could simplify the code and gain a few cycles back:

```brightscript
'Available since v10.5
if m.memMonitor <> invalid
    ev.AddReplace("memLimitPercent", m.memMonitor.GetMemoryLimitPercent())
    'Available for RokuOS v12.5+
    if m.enableChannelAvailableMem
        ev.AddReplace("channelAvailMem", m.memMonitor.GetChannelAvailableMemory())
    end if
end if
```

Reply:
It makes sense.

Reply:
This solution sadly doesn't work. The `roAppMemoryMonitor` object can't be stored in "m".

Reply:
Unfortunately, this solution is not possible because of how roku handles the references to some objects across thread boundaries. If I create a `roAppMemoryMonitor` object during the init and store it in "m", this object will only exist in the same thread the init was called (the thread where the `NewRelic()` constructor is called). Once you pass the NR Agent reference to, let's say, a `Screne` where the video player resides, this will be a different thread, and the reference to the `roAppMemoryMonitor` will become invalid.

Reply:
Fair enough. But we could still at least optimise it to the point where most checks are done on init. So `roAppMemoryMonitor` won't get recycled but we avoid the cost of checking for device strings and toggling memory warning events.

```brightscript
' on init

m.enableMemMonitor = false
m.enableChannelAvailableMem = false

if isMemoryMonitorAvailable(dev.GetModel())
    m.enableMemMonitor = true

    'Available since v10.5
    memMonitor = CreateObject("roAppMemoryMonitor")
    if memMonitor <> invalid
        'UNDOCUMENTED TRICK:
        '   If EnableMemoryWarningEvent(true) returns false, a call to GetChannelAvailableMemory will freeze the device on v12.5
        enableMemWarningRet = memMonitor.EnableMemoryWarningEvent(true)
        memMonitor.EnableMemoryWarningEvent(false)

        if enableMemWarningRet
            'Available for RokuOS v12.5+
            m.enableChannelAvailableMem = FindMemberFunction(memMonitor, "GetChannelAvailableMemory") <> Invalid
        end if
    end if
end if

' on nrAddAttributes()

if m.enableMemMonitor <> invalid
    memMonitor = CreateObject("roAppMemoryMonitor")
    ev.AddReplace("memLimitPercent", memMonitor.GetMemoryLimitPercent())
    'Available for RokuOS v12.5+
    if m.enableChannelAvailableMem
        ev.AddReplace("channelAvailMem", m.memMonitor.GetChannelAvailableMemory())
    end if
end if
```

Reply:
This solution works, I'm going to apply it, thanks for your suggestions. One last note, I found out that, when `GetChannelAvailableMemory()` is not available, `GetMemoryLimitPercent()` always returns 0, so I'm simplifying the logic and just have one flag to control the execution of both functions.